src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -316,9 +316,30 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 self.write_scalar(offset_ptr, dest)?;
             }
             sym::ptr_guaranteed_eq | sym::ptr_guaranteed_ne => {
-                // FIXME: return `true` for at least some comparisons where we can reliably
-                // determine the result of runtime (in)equality tests at compile-time.
-                self.write_scalar(Scalar::from_bool(false), dest)?;
+                let a = self.read_immediate(args[0])?.to_scalar()?;
+                let b = self.read_immediate(args[1])?.to_scalar()?;
+                // `Some` in `cmp` means the comparison is known, guaranteed in both directions
+                let cmp = match (a, b) {
+                    // Comparisons between integers are always known.
+                    (Scalar::Raw { .. }, Scalar::Raw { .. }) => Some(a == b),
+                    // Comparisons with null pointers are always known, no abstract pointer can be
+                    // equal to the null pointer,
```

Comment:
That's not true. For every `i: usize` and `alloc_id: AllocId` there exists exactly one `offset: usize` such that `Pointer { alloc_id, offset }` is equal to `i`. This holds even for `i == 0`. NULL is special only for dereferencability, not for equality.

Reply:
The only time we can be sure that a pointer is not NULL is when it is dereferencable. Also see [`ptr_may_be_null`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/struct.Memory.html#method.ptr_may_be_null) (the negation of which is `guaranteed_ne` for NULL).

Reply:
oh right... I was only thinking about base pointers. Who ever thought offsetting pointers was a good idea? :laughing: